### DISCUSSION TITLE
Strongswan not decrypting unencapsulated ESP

### DISCUSSION THREAD

**Discussion by ptchuba** [strongswan/strongswan#1316] [Q&A]
My setup is as follows.

I'm using Ikev2 and IPv6 for a roadwarrior connection. Outer transport is IPv6. Assigning both IPv4 and IPv6 addresses inside tunnel. I'm using machine certificates for authentication. 

Using strongswan on linux client  works fine. Connection to VPN is successful and I have both inner IPv4 and IPv6 reachability. (Note that linux client is on Azure. Uses ULA address which gets natted to Public IPv6).

Using the strongswan app on Android works fine as well. However, the native app does not connect (problem for a different day)

My main issue is that a Windows client connects fine but won't pass traffic. I get ipv4 and ipv6 addresses but can pass neither traffic through tunnel. For testing purposes, I enabled IPv4 as the outer transport and can successfully connect and pass v4 and v6 traffic through tunnel. In my tests, I noticed that when connecting to gateway using IPv6, my tunnel bound traffic is ESP but NOT encapsulated in UDP as is the case with connecting to gateway using IPv4. i.e. UDP 4500 is not used to encapsulate ESP when connecting using IPv6. 

I have ensured that my strongswan server is receiving plain ESP traffic. strongswan server is on Azure so public IPv6 is not on the vm itself. Had to use cli commands to add ESP protocol 50 to network security group. TCPdump on server shows ESP traffic from windows client. See below.

``` 	
00:07:04.673748 IP6 (hlim 113, next-header ESP (50) payload length: 116) 2601:xxx:8201:14c0::6ddf > fd01:1:2:1::4: ESP(spi=0xcd5831e2,seq=0x1), length 116
00:07:09.409642 IP6 (hlim 113, next-header ESP (50) payload length: 116) 2601:xxx:8201:14c0::6ddf > fd01:1:2:1::4: ESP(spi=0xcd5831e2,seq=0x2), length 116
00:07:14.405106 IP6 (hlim 113, next-header ESP (50) payload length: 116) 2601:xxx:8201:14c0::6ddf > fd01:1:2:1::4: ESP(spi=0xcd5831e2,seq=0x3), length 116
00:07:19.419120 IP6 (hlim 113, next-header ESP (50) payload length: 116) 2601:xxx:8201:14c0::6ddf > fd01:1:2:1::4: ESP(spi=0xcd5831e2,seq=0x4), length 116
``` 

Wireshark from client looks like this:
<img width="1175" alt="image" src="https://user-images.githubusercontent.com/6981814/192401965-608c4eb1-fb79-4426-af5e-e20a9d2fc0a6.png">

Conversely, if I used IPv4 to connect to the strongswan server, wireshark looks like this.
<img width="875" alt="image" src="https://user-images.githubusercontent.com/6981814/192400552-38d715b3-02f2-48c1-9b6d-87c6e8c0d1ac.png">

Here is my strongswan roadwarrior connection snippet. Note that changing to "encap =no" seems to have no effect, at least on the Windows side. It still sends out ESP unecapsulated.

```
 rw {
      local_addrs  = %any
      pools = rw_pool_v4, rw_pool_v6
      send_cert = always
      local {
         auth = pubkey
         certs = vdi-ep1.pem
         id = vdi-ep1.example.local
      }
      remote {
         auth = pubkey
      }
      children {
         client-sa {
            local_ts  = fd01:1:5:2::4/64,fd02:1:1:1::eeee/128,172.25.0.254,10.6.1.0/24
            if_id_out = 2
            if_id_in = 2
            esp_proposals = aes256gcm128-ecp384
            mode = tunnel
         }
      }
      version = 2
      encap = yes
      proposals = aes256gcm16-prfsha512-ecp384,aes256gcm16-prfsha384-ecp384
   }
```

And this is the ipsec statusall SA. Notice the line "ESP in UDP SPIs".

```
Security Associations (3 up, 0 connecting):
          rw[4]: ESTABLISHED 7 seconds ago, fd01:1:2:1::4[vdi-ep1.example.local]...2601:xxx:8201:14c0::6ddf[CN=strongswan-vdi2, C=US, L=Austin, O=example, OU=IT]
          rw[4]: IKEv2 SPIs: 77c948761f1f86b1_i 03c418f44134a4ca_r*, rekeying in 3 hours
          rw[4]: IKE proposal: AES_GCM_16_256/PRF_HMAC_SHA2_384/ECP_384
       rw-sa{6}:  INSTALLED, TUNNEL, reqid 6, ESP in UDP SPIs: cd5831e2_i 52f6647c_o
       rw-sa{6}:  AES_GCM_16_256, 0 bytes_i, 0 bytes_o, rekeying in 56 minutes
       rw-sa{6}:   10.6.1.0/24 172.25.0.254/32 fd01:1:5:2::/64 fd02:1:1:1::eeee/128 === 172.25.0.1/32 fd02:1:1:1::2/128
```
Notice the line "ESP in UDP SPIs" above. To me this means the server side of the tunnel is expecting ESP packets encapsulated in UDP (and rightfully so since it is behind a NAT) and so is not decrypting the native ESP packets.

If I'm right I think i have 2 options:
1. Force Windows to encapsulate IPv6 ESP in UDP - Could not figure out how to do this. Tried adding `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PolicyAgent
AssumeUDPEncapsulationContextOnSendRule = 2` with no luck.
2. Somehow convince strongswan to decrypt native ESP packets with same spi - no clue how to start.

Any and all help will be greatly appreciated.

**Comment by tobiasbrunner** [comment]
> Force Windows to encapsulate IPv6 ESP in UDP

It's definitely possible that Windows doesn't support UDP-encapsulation for IPv6. The Linux kernel, for instance, only can do that since 5.8, which was released just a couple of years ago.

> Somehow convince strongswan to decrypt native ESP packets with same spi - no clue how to start.

That's not a strongSwan problem as traffic is handle by the Linux kernel.  However, the kernel currently doesn't support processing plain ESP packets if the IPsec SA has UDP-encapsulation enabled (and vice-versa). The latter happens automatically if a NAT is detected. So to avoid it, you'd have to prevent NAT detection. There is currently no option to disable that in strongSwan, though. Therefore, if you wanted to try that, you'd have to patch the _ike-natd_ task manually. The simplest would probably be to remove this line: https://github.com/strongswan/strongswan/blob/ef93c7e2ea13ce0b90dcc0dfb3316a3115c830f0/src/libcharon/sa/ikev2/tasks/ike_natd.c#L480

However, that disables it completely for both IPv4 and IPv6 and as initiator and responder. So if you can't go back to using IPv4, there is currently no simple fix.

**Reply by ptchuba** [reply]
Thanks @tobiasbrunner. Greatly appreciate you taking a look at this for me. Seems I have no way out here. Its crazy we still can't get stuff to work universally with IPv6 in 2022. Will bring this up to Microsoft's attention though I doubt anything will be done soon if at all.

**Comment by geppi** [comment]
Sorry to capture this thread but I stumbled across it when researching a VPN interoperability problem between StrongSwan and a Lancom router.
@tobiasbrunner You are saying that:
> the kernel currently doesn't support processing plain ESP packets if the IPsec SA has UDP-encapsulation enabled (and vice-versa)

This was exactly the problem encountered with the Lancom router, where after IKE negotiation the StrongSwan VPN client did send plain vanilla ESP pakets while the Lancom router did answer with UDP encapsulated ESP. The Lancom router obviously did process plain ESP packets although it did use UDP encapsulation itself. BTW, the solution for the problem was to force UDP encapsulation in the StrongSwan client.
However, according to the responsible developer at Lancom, their implementation is compliant with [RFC 5996](https://www.rfc-editor.org/rfc/rfc5996) :

> Port 4500 is reserved for UDP-encapsulated ESP and IKE.  An IPsec
   endpoint that discovers a NAT between it and its correspondent (as
   described below) MUST send all subsequent traffic from port 4500,
   which NATs should not treat specially (as they might with port 500).

  > An initiator can use port 4500 for both IKE and ESP, regardless of
   whether or not there is a NAT, even at the beginning of IKE.  When
   either side is using port 4500, sending ESP with UDP encapsulation is
   not required, but understanding received UDP-encapsulated ESP packets
   is required.  UDP encapsulation MUST NOT be done on port 500.  If
   Network Address Translation Traversal (NAT-T) is supported (that is,
   if NAT_DETECTION_*_IP payloads were exchanged during IKE_SA_INIT),
   all devices MUST be able to receive and process both UDP-encapsulated
   ESP and non-UDP-encapsulated ESP packets at any time.  Either side
   can decide whether or not to use UDP encapsulation for ESP
   irrespective of the choice made by the other side.  However, if a NAT
   is detected, both devices MUST use UDP encapsulation for ESP.

Does that mean that the current implementation of IKEv2 in the Linux kernel is not fully compliant with RFC 5996?

**Reply by tobiasbrunner** [reply]
> However, according to the responsible developer at Lancom, their implementation is compliant with [RFC 5996](https://www.rfc-editor.org/rfc/rfc5996)

Technically, sure. But did they give a reason why the box used UDP-encapsulation if there was no NAT detected? (By the way, that RFC was replaced by RFC 7296. Not that it makes a difference in this case.)

> Does that mean that the current implementation of IKEv2 in the Linux kernel is not fully compliant with RFC 5996?

Not the IKEv2 implementation (which is strongSwan), but the IPsec/ESP implementation in the kernel is the issue.

**Comment by geppi** [comment]
> did they give a reason why the box used UDP-encapsulation if there was no NAT detected?

Not really. They simply say that the router behaves compliant with the RFC.
If NAT detection is used and the client at least once uses port 4500 during the IKE negotiation process they seem to send UDP encapsulated ESP, which is also an issue for the native Windows 10/11 VPN client because it also doesn't process a mixed UDP enc. ESP/plain ESP communication.

Fortunately there exist workarounds in both cases, i.e. the mentioned "force UDP enc." setting in StrongSwan and disabling MOBIKE in the Windows VPN client.

I was just curious if the IPsec/ESP implementation in the Linux kernel is really not RFC compliant.

However, good that StrongSwan has the flexibility to force the UDP encapsulation.
It looks like Lancom is reluctant to at least implement a switch to disable the UDP encapsulation.

**Reply by tobiasbrunner** [reply]
> They simply say that the router behaves compliant with the RFC.
> If NAT detection is used and the client at least once uses port 4500 during the IKE negotiation process they seem to send UDP encapsulated ESP,

Hm, using the port as trigger doesn't really make sense because a switch to port 4500 is very common due to MOBIKE (and since there is an actual mechanism to detect NATs already built into IKEv2). And while it is compliant with the RFC to send UDP-encapsulated ESP packets even if no NAT was detected, it usually doesn't make sense to add that overhead if it's not necessary (in particular if it's known that it causes interoperability issues with other implementations). I wonder if they'd use UDP-encap too if IPv6 was used and there is a switch to port 4500 (only newer Linux kernels even support that, although again, only if UDP-encap is explicitly enabled).

**Comment by geppi** [comment]
Sorry I didn't provide this detail but it's all about IPv6. With IPv4 it's much less of a problem due to the ubiquitous NAT where UDP enc. is used anyhow.
No, they really use UDP enc. with IPv6 as soon as there was a switch to port 4500 in the negotiation process. As you said it's very common that this happens due to MOBIKE and it is the reason why disabling MOBIKE is a workaround for the native Windows VPN client. The latter switches to port 4500 when MOBIKE is enabled. However, the Lancom router doesn't support MOBIKE, so it looks like the Windows client decides that UDP enc. is unecessary and configures the SA without it. Issuing a  '_Get-NetIPsecMainModeSA -All_' on the Windows client shows empty lines for '_LocalUdpEncapsulationPort_' as well as '_RemoteUdpEncapsulationPort_'. I don't know if the Windows kernel has the same limitation as the Linux kernel with respect to a "mixed UDP enc. mode", in any case it's not processing the UDP enc. ESP packets from the Lancom router.

The fact that the router doesn't support MOBIKE makes the choice of UDP enc. with IPv6 even more unecessary, right?

**Reply by tobiasbrunner** [reply]
> Sorry I didn't provide this detail but it's all about IPv6. With IPv4 it's much less of a problem due to the ubiquitous NAT where UDP enc. is used anyhow.

Ah, thanks for the clarification.

> No, they really use UDP enc. with IPv6 as soon as there was a switch to port 4500 in the negotiation process.

OK, but then it makes even less sense. One of the advantages of using IPv6 is avoiding all those NAT/encap shenanigans. And again, UDP-encap for IPv6 only works with newer Linux kernels (at least 5.8, which was released in 2020, so I guess it's not that new anymore :-).

> The latter switches to port 4500 when MOBIKE is enabled.

Yes, strongSwan does the same as it's [required by MOBIKE](https://datatracker.ietf.org/doc/html/rfc4555#section-3.3).

> I don't know if the Windows kernel has the same limitation as the Linux kernel with respect to a "mixed UDP enc. mode", in any case it's not processing the UDP enc. ESP packets from the Lancom router.

Sounds like it. If we don't detect a NAT, we also don't set any encap mode or ports on the SAs, which then causes the Linux kernel to drop encapsulated packets (`XfrmInStateMismatch` should increase in `/proc/net/xfrm_stat`), which it obviously shouldn't. On the other hand, it also drops plain ESP packets if the SA has encapsulation enabled.

While some of this can probably be changed quite easily, there are aspects that could be tricky. For instance, the kernel supports various types of hardware offloading and tunneling interfaces and some of these work with static offsets that are pre-calculated per SA to find the start of the ESP header, which wouldn't work anymore (either too much or too little would be skipped).

> The fact that the router doesn't support MOBIKE makes the choice of UDP enc. with IPv6 even more unecessary, right?

UDP-encap doesn't really have anything to do with MOBIKE. It works perfectly fine between e.g. two multi-homed hosts that aren't behind any NATs. The port is only changed to the "NAT-T port" to avoid possibly having to switch it later if a client moves behind a NAT. As specified in the text you referenced earlier, IKEv2 allows using port 4500 from the start anyway (i.e. even before it's known whether there is a NAT or not).

**Comment by geppi** [comment]
BTW, is there a possibility to disable MOBIKE when using the StrongSwan NetworkManager plugin?

If I enter _'mobike=no'_ into the NetworkManager configuration file _'/etc/NetworkManager/system-connections/\<conn\>.nmconnection'_ it gets overwritten every time I change the config in the GUI and I'm not sure if this option is recognized in the NetworkManager configuration file at all.

Currently I'm using the workaround of enforcing UDP enc. with StrongSwan to make the VPN-Tunnel built over IPv6 work with the Lancom router, because this option is available in the GUI. However, following our discussion above I would prefer to use the same workaround as with the Windows VPN client to avoid UDP enc. at all.

**Reply by tobiasbrunner** [reply]
No, MOBIKE currently can't be disabled in the NM plugin (the extension will obviously be disabled because the responder doesn't support it, but at that point the port has already been switched because that happens at the same time the notify is sent that announces support for the feature). Again, using the port switch as an indicator to use UDP-encap seems like a Lancom issue (while not technically in violation of the RFC, it unnecessarily provokes interoperability issues).

> If I enter _'mobike=no'_ into the NetworkManager configuration file _'/etc/NetworkManager/system-connections/<conn>.nmconnection'_ it gets overwritten every time I change the config in the GUI and I'm not sure if this option is recognized in the NetworkManager configuration file at all.

That config file has nothing to do with swanctl.conf/ipsec.conf so you can't just set arbitrary options there.

This would require a new option similar to the force encap one (GUI and backend). Please go ahead and file a feature request if you think that would be useful.

**Comment by geppi** [comment]
Thank you very much for taking the time to look into this.

Since there is already a possibility to deal with this Lancom router peculiarity in form of the UDP enc. enforcement I don't want to waste anybody's time to implement a feature that probably nobody else would require.

However, if you think that disabling MOBIKE in the NetworkManager plugin could be of use in other cases I'm happy to open a feature request based on my use case.

**Reply by tobiasbrunner** [reply]
Yeah, I currently don't really see a reason to provide such an option, given that the NM plugin is usually used in roadwarrior scenarios where mobility is often a core aspect.
